napari/utils/migrations.py

```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
     _OldClass.__new__.__signature__ = prealloc_signature  # type: ignore [attr-defined]
 
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
+    """A dictionary that issues warning messages when deprecated keys are accessed.
+
+    Deprecated keys and values are not stored as part of the dictionary, so will not
+    appear when iterating over this or its items.
+
+    Instead deprecated items can only be accessed using `__getitem__`, `__setitem__`,
+    and `__delitem__`, or using `self.deprecations` directly.
+    """
+
+    # Maps from a deprecated key to its value and deprecation message.
+    deprecations: dict[str, tuple[Any, str]]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.deprecations = {}
```

Comment:
```suggestion
        self._deprecations = {}
```
I think that this should be private value

Reply:
It might actually be a useful interface for users to inspect things...

Reply:
I'm a little undecided here.

I like this being public because it helps inspection/explanation and also gives simple flexibility on how things are deprecated.

But I like this being private because I don't want people adding arbitrary items to this dict.

Let me try making it private and adding some public helpers.

Reply:
I made this private, added two separate public methods to add/set deprecated items, then added `deprecated_keys` as a public property. I think that works well enough.